File: tests/acceptance/tests/Performance/generate_separate_categories.py
import csv
import uuid
import random
import argparse
import urllib.parse
from datetime import datetime
from faker import Faker
from tqdm import tqdm

fake = Faker()

# === 🔧 Column Toggle Config ===
ENABLED_COLUMNS = {
    "id": True,
    "parent_id": True,
    "active": True,
    "type": True,
    "visible": True,
    "name": True,
    "external_link": True,
    "description": True,
    "meta_title": True,
    "meta_description": True,
}

# === 🧭 Top-Level Parent ID (for all mid categories)
TOP_PARENT_ID = "0196156e3ecf729ea8b737e4911e99b9"

# === Storage
MID_CATEGORIES = {}      # mid_id -> dict
SUB_CATEGORIES = {}      # mid_id -> list[dict]


def maybe(prob, generator):
    """Randomly return empty string or value from generator."""
    return generator() if random.random() > prob else ""


def generate_slug_from_name(name: str) -> str:
    return name.lower().replace(" ", "-")


def generate_url_from_slug(slug: str) -> str:
    return f"https://shop.example.com/collections/{urllib.parse.quote(slug)}"


def generate_category_tree(mid_count, sub_per_mid):
    MID_CATEGORIES.clear()
    SUB_CATEGORIES.clear()

    for _ in range(mid_count):
        mid_id = uuid.uuid4().hex
        mid_name = fake.bs().title()
        MID_CATEGORIES[mid_id] = {
            "id": mid_id,
            "name": mid_name,
            "parent_id": TOP_PARENT_ID
        }

        SUB_CATEGORIES[mid_id] = []
        for _ in range(sub_per_mid):
            sub_id = uuid.uuid4().hex
            sub_name = fake.bs().title()
            SUB_CATEGORIES[mid_id].append({
                "id": sub_id,
                "name": sub_name,
                "parent_id": mid_id
            })


def build_row(cat_id, parent_id, name_prefix, name, null_prob):
    full_name = f"{name_prefix}: {name}"
    slug = generate_slug_from_name(name)
    url = generate_url_from_slug(slug)

    row = {
        "id": cat_id,
        "parent_id": parent_id,
        "active": 1,  # or random.choice([0, 1])
        "type": "page",
        "visible": 1,
        "name": full_name,
        "external_link": maybe(null_prob, lambda: fake.url()),
        "description": fake.paragraph(nb_sentences=3),
        "meta_title": maybe(null_prob, lambda: fake.sentence(nb_words=6)),
        "meta_description": maybe(null_prob, lambda: fake.text(max_nb_chars=100)),
    }

    return [row[col] for col in ENABLED_COLUMNS if ENABLED_COLUMNS[col]]


def write_csv(path, rows, null_prob, delimiter, include_header, label):
    headers = [col for col, enabled in ENABLED_COLUMNS.items() if enabled]

    with open(path, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.writer(csvfile, delimiter=delimiter, quotechar='"', quoting=csv.QUOTE_MINIMAL)

        if include_header:
            writer.writerow(headers)

        with tqdm(total=len(rows), desc=f"Writing {label}", unit="row") as pbar:
            for row in rows:
                writer.writerow(row)
                pbar.update(1)


def generate_csv_files(mid_out, sub_out, delimiter, null_prob, include_header):
    mid_rows = []
    sub_rows = []

    for mid_id, mid in MID_CATEGORIES.items():
        mid_rows.append(build_row(
            cat_id=mid["id"],
            parent_id=mid["parent_id"],
            name_prefix="Mid",
            name=mid["name"],
            null_prob=null_prob
        ))

    for mid_id, sub_list in SUB_CATEGORIES.items():
        for sub in sub_list:
            sub_rows.append(build_row(
                cat_id=sub["id"],
                parent_id=sub["parent_id"],
                name_prefix="Sub",
                name=sub["name"],
                null_prob=null_prob
            ))

    write_csv(mid_out, mid_rows, null_prob, delimiter, include_header, "Mid Categories")
    write_csv(sub_out, sub_rows, null_prob, delimiter, include_header, "Sub Categories")

    print(f"\n✅ Mid categories: {len(mid_rows)} → {mid_out}")
    print(f"✅ Sub categories: {len(sub_rows)} → {sub_out}")


def parse_args():
    # Generate timestamp for default filenames
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    
    parser = argparse.ArgumentParser(description="Generate Shopify-style category hierarchy into separate CSVs.")
    parser.add_argument("--mid-count", type=int, default=50, help="Number of mid-level categories")
    parser.add_argument("--sub-per-mid", type=int, default=100, help="Number of subcategories per mid")
    parser.add_argument("--mid-out", type=str, default=f"mid_categories_{timestamp}.csv", help="Output file for mid categories")
    parser.add_argument("--sub-out", type=str, default=f"sub_categories_{timestamp}.csv", help="Output file for sub categories")
    parser.add_argument("--delimiter", type=str, default=",", help="CSV delimiter")
    parser.add_argument("--null-prob", type=float, default=0.0, help="Probability that optional fields are empty")
    parser.add_argument("--no-header", action="store_true", help="Omit CSV header row")
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()

    generate_category_tree(mid_count=args.mid_count, sub_per_mid=args.sub_per_mid)

    generate_csv_files(
        mid_out=args.mid_out,
        sub_out=args.sub_out,
        delimiter=args.delimiter,
        null_prob=args.null_prob,
        include_header=not args.no_header
    )

File: tests/acceptance/tests/Performance/generate_separate_categories.js
// generate_separate_categories_dual.js
import fs from 'fs';
import { faker } from '@faker-js/faker';
import { argv } from 'process';

// === Config ===
const ENABLED_COLUMNS = {
    id: true,
    parent_id: true,
    active: true,
    type: true,
    visible: true,
    name: true,
    external_link: true,
    description: true,
    meta_title: true,
    meta_description: true,
    products: true
};

const TOP_PARENT_ID = '01983b11d12e71d1859626764f252977';
const PRODUCT_ID = '01983b16e4667063a3fb8ae4b2539715';

function maybe(prob, generator) {
    return Math.random() > prob ? generator() : '';
}

function generateSlugFromName(name) {
    return name.toLowerCase().replace(/\s+/g, '-');
}

function generateUrlFromSlug(slug) {
    return `https://shop.example.com/collections/${encodeURIComponent(slug)}`;
}

function buildRow(cat, nullProb) {
    const slug = generateSlugFromName(cat.name);
    const fullName = `${cat.prefix}: ${cat.name}`;

    const row = {
        id: cat.id,
        parent_id: cat.parent_id,
        active: faker.number.int({ min: 0, max: 1 }),
        type: faker.helpers.arrayElement(['page', 'folder', 'link']),
        visible: faker.number.int({ min: 0, max: 1 }),
        name: fullName,
        external_link: maybe(nullProb, () => faker.internet.url()),
        description: faker.lorem.paragraph(3),
        meta_title: maybe(nullProb, () => faker.lorem.sentence(6)),
        meta_description: maybe(nullProb, () => faker.lorem.text().substring(0, 100)),
        products: PRODUCT_ID
    };

    return Object.entries(ENABLED_COLUMNS)
        .filter(([_, enabled]) => enabled)
        .map(([col]) => row[col]);
}

function parseArgs() {
    const args = {};
    for (let i = 2; i < argv.length; i++) {
        if (argv[i].startsWith('--')) {
            const key = argv[i].replace(/^--/, '');
            const value = argv[i + 1] && !argv[i + 1].startsWith('--') ? argv[++i] : true;
            args[key] = value;
        }
    }
    return {
        midCount: parseInt(args['mid-count'] ?? '50', 10),
        subPerMid: parseInt(args['sub-per-mid'] ?? '100', 10),
        midOut: args['mid-out'] ?? `mid_categories_${Date.now()}.csv`,
        subOut: args['sub-out'] ?? `sub_categories_${Date.now()}.csv`,
        delimiter: args['delimiter'] ?? ',',
        nullProb: parseFloat(args['null-prob'] ?? '0'),
        noHeader: !!args['no-header']
    };
}

function writeDualCSVStream({ midCount, subPerMid, midOut, subOut, delimiter, nullProb, noHeader }) {
    const midStream = fs.createWriteStream(midOut, { encoding: 'utf-8' });
    const subStream = fs.createWriteStream(subOut, { encoding: 'utf-8' });
    const headers = Object.entries(ENABLED_COLUMNS)
        .filter(([_, enabled]) => enabled)
        .map(([col]) => col);

    if (!noHeader) {
        midStream.write(headers.join(delimiter) + '\n');
        subStream.write(headers.join(delimiter) + '\n');
    }

    let midCountWritten = 0;
    let subCountWritten = 0;

    for (let i = 0; i < midCount; i++) {
        // Mid category
        const midId = faker.string.uuid().replace(/-/g, '');
        const midName = faker.commerce.department();
        const midRow = buildRow({ id: midId, parent_id: TOP_PARENT_ID, prefix: 'Mid', name: midName }, nullProb);
        midStream.write(midRow.join(delimiter) + '\n');
        midCountWritten++;

        // Sub categories
        for (let j = 0; j < subPerMid; j++) {
            const subId = faker.string.uuid().replace(/-/g, '');
            const subName = faker.commerce.department();
            const subRow = buildRow({ id: subId, parent_id: midId, prefix: 'Sub', name: subName }, nullProb);
            subStream.write(subRow.join(delimiter) + '\n');
            subCountWritten++;
        }

        if (i % 10 === 0) {
            process.stdout.write(`\rMid: ${midCountWritten} | Sub: ${subCountWritten} rows written...`);
        }
    }

    midStream.end();
    subStream.end(() => {
        console.log(`\n✅ Done! ${midCountWritten} mid categories → ${midOut}`);
        console.log(`✅ Done! ${subCountWritten} sub categories → ${subOut}`);
    });
}

if (import.meta.url === `file://${process.argv[1]}`) {
    const args = parseArgs();
    writeDualCSVStream(args);
}
